### PR TITLE
Fix decorator generation

### PR DESCRIPTION
**Pull request by PiotrMachner**
- support for decorating non-hierarchical types with `class-transfomer`'s `@Type` as well (it is needed for `class-validator` to recognise all types correctly, which is mandatory when using the `forbidUnknownValues` flag as recommended by Snyk)
- support for decorating array/collection fields with `class-transfomer`'s `@Type` (for the same reason)
- adding the `@IsOptional` decorator where needed with `@MaxLength` / `@MinLength` (which by default don't accept null/undefined)

### CODE REVIEW

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java [hunk 1]
```diff
@@ -59,29 +57,43 @@ private TsBeanModel decorateClass(SymbolTable symbolTable, TsBeanModel bean) {
...
     }
 
+    private Class<?> resolveTypeToDecorate(ParameterizedTypeClasses parameterizedTypeClasses) {
+        Class<?> type = parameterizedTypeClasses.getType();
+        Class<?> typeToDecorate = type;
+        if (type.isArray()) {
+            typeToDecorate = type.getComponentType();
+        }
+
+        final Optional<Class<?>> typeArgument = parameterizedTypeClasses.getTypeArgument();
```
Comment (lpandzic) [inline infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: final doesn't really bring any kind of value here

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java [hunk 2]
```diff
@@ -98,6 +110,20 @@ private TsPropertyModel getDecorators(SymbolTable symbolTable, TsBeanModel model
                      .collect(Collectors.toList());
     }
 
+    private List<TsDecorator> getNonHierarchyDecorators(TsPropertyModel tsPropertyModel,
+                                                        Class<?> type) {
+
+        TsArrowFunction emptyToTypeName = new TsArrowFunction(Collections.emptyList(), new TsTypeReferenceExpression(
+                new TsType.ReferenceType(new Symbol(type.getSimpleName()))));
+
+        final Stream<TsDecorator> typeDecoratorStream = shouldNotBeDecorated(type) ?
```
Comment (lpandzic) [inline infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: same as before

Comment (lpandzic) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: instead of ternary operator with extracted flag method why not a extracted method that uses guard clauses and returns the target type (Stream)?

Comment (lpandzic) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: instead of 3 boolean methods you can have multiple guard clauses that return Stream.empty on each separate condition

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java [hunk 3]
```diff
@@ -133,4 +171,19 @@ private void appendToParentToChildren(Class<?> key, Stream<? extends Class<?>> v
             return Optional.empty();
         }
     }
+
+    private ParameterizedTypeClasses getParameterizedTypeClasses(Field field) {
+        return new ParameterizedTypeClasses(field.getType(), getTypeArgument(field));
+    }
+
+    private Optional<Class<?>> getTypeArgument(Field field) {
+        final Type genericType = field.getGenericType();
```
Comment (lpandzic) [inline infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: final

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/validation/SizeToTsDecoratorConverter.java [hunk 4]
```diff
@@ -21,12 +20,14 @@ public SizeToTsDecoratorConverter(ValidationMessageReferenceResolver validationM
     public Stream<TsDecorator> convert(Field field, Size annotation) {
         String min = String.valueOf(annotation.min());
         if (field.getType().equals(String.class)) {
-            return Stream.concat(max("MaxLength", annotation, annotation.max()),
-                                 getMin(field, annotation, min));
+            return Stream.of(getMax("MaxLength", annotation, annotation.max()),
```
Comment (msertic1) [inline infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/validation/SizeToTsDecoratorConverter.java:23]: just a question, so with this piece of logic we are assuring that if field is annotated with @Size in java world and there is no annotations: NotBlank.class, NotNull.class, NotEmpty.class we are going to annotate it additionally with @IsOptional when transferring to TS world ?

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java [hunk 5]
```diff
@@ -114,6 +140,18 @@ private TsPropertyModel getDecorators(SymbolTable symbolTable, TsBeanModel model
...
+    private boolean shouldNotBeDecorated(Class<?> type) {
+        return isBuiltIn(type) || isTsTypeResolutionUnsupported(type);
+    }
+
+    private boolean isTsTypeResolutionUnsupported(Class<?> type) {
+        return type.isEnum() || PresentPropertyJsonHierarchy.class.isAssignableFrom(type);
+    }
+
+    private boolean isBuiltIn(Class<?> type) {
+        return Optional.ofNullable(type.getPackage()).map(Package::getName).orElse("").startsWith("java");
```
Comment (msertic1) [inline infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java:160]: Question:
We are skipping decoration for built in java types, if so i would rename method to isBuiltInType

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java [hunk 6]
```diff
@@ -114,6 +140,18 @@ private TsPropertyModel getDecorators(SymbolTable symbolTable, TsBeanModel model
                        .collect(Collectors.toList());
     }
 
+    private boolean shouldNotBeDecorated(Class<?> type) {
+        return isBuiltIn(type) || isTsTypeResolutionUnsupported(type);
+    }
+
+    private boolean isTsTypeResolutionUnsupported(Class<?> type) {
+        return type.isEnum() || PresentPropertyJsonHierarchy.class.isAssignableFrom(type);
```
Comment (msertic1) [inline infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java:156]: Could you explain it to me why is resolution is supported if -> PresentPropertyJsonHierarchy.class.isAssignableFrom(type);

It is just question from my side to be able to better understand this feature :)

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java [hunk 7]
```diff
@@ -59,29 +57,43 @@ private TsBeanModel decorateClass(SymbolTable symbolTable, TsBeanModel bean) {
...
+        return typeToDecorate;
+    }
+
     private List<TsDecorator> getDecorators(SymbolTable symbolTable,
                                             TsBeanModel model,
                                             TsPropertyModel tsPropertyModel,
                                             Class<?> type) {
         return factory.create(type)
                       .filter(resolver -> resolver instanceof CompositeJsonTypeResolver<?>)
+                      .filter(resolver -> !Modifier.isFinal(type.getModifiers()))
```
Comment (msertic1) [inline infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: why we have added this filter ?

**Review by lpandzic — CHANGES_REQUESTED** [review]

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java (hunk 7 above)
Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: As you see below, these filters determine whether there will be a hierarchical `@Type` decorator like
```
@Type(() => Object, {
        discriminator: {
            property: 'type', subTypes: [
                { value: ReplySuggestion, name: SuggestionType.REPLY },
                { value: OpenUrlSuggestion, name: SuggestionType.OPEN_URL }
            ]
        }
    })
```
or a 'flat' one like this: `@Type(() => CardContent)`

I had a case where I had an array of objects of a type that extended `SimpleJsonHierarchy` so the above filter `resolver instanceof CompositeJsonTypeResolver<?>` returned true, but the type itself was final (it was a hierarchy 'leaf') so I just wanted the 'flat' decorator.

I'll do some refactoring here to improve readability.

Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: It's refactored now and hopefully my intensions are now more clear :)

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java (hunk 6 above)
Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java:156]: This is just because resolution of hierarchies implemented using `PresentPropertyJsonHierarchy` has not been needed so far and is not implemented yet

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java (hunk 5 above)
Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java:160]: Done

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/validation/SizeToTsDecoratorConverter.java (hunk 4 above)
Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/validation/SizeToTsDecoratorConverter.java:23]: Exactly. This is needed because Java `@Size` passes validation for nulls, but Typescript `@MaxLength` and `@MinLength` do not allow null or undefined.

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java (hunk 3 above)
Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: Removed

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java (hunk 1 above)
Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: Removed

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java (hunk 2 above)
Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: `final` removed

**Review by lpandzic — DISMISSED** [review]

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java (hunk 7 above)
Comment (msertic1) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java]: double check if this  !Modifier.isFinal(type.getModifiers()) is good enough and if it covers all cases

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java [hunk 8]
```diff
@@ -59,29 +57,50 @@ private TsBeanModel decorateClass(SymbolTable symbolTable, TsBeanModel bean) {
...
+        return typeToDecorate;
+    }
+
     private List<TsDecorator> getDecorators(SymbolTable symbolTable,
                                             TsBeanModel model,
                                             TsPropertyModel tsPropertyModel,
                                             Class<?> type) {
         return factory.create(type)
-                      .filter(resolver -> resolver instanceof CompositeJsonTypeResolver<?>)
+                      .filter(resolver -> isHierarchicalDecoratorNeeded(resolver, type))
```
Comment (msertic1) [inline infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java:86]: oh i get it, you have reverse logic, instead of finding leaves you have detected root :) nice, i think that we have now better use case coverage :)

File: infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java [hunk 9]
```diff
@@ -9,10 +9,13 @@
 
 abstract class ClassValidatorDecoratorTestBase extends TestBase {
 
+    private static final String IMPORT_COMMON_VALIDATION_MESSAGES = "import { CommonValidationMessages } from 'infobip-typescript-generator-common'";
+    private static final String IMPORT_CLASS_VALIDATOR = "import { ValidateNested, IsDefined, IsOptional, IsNotEmpty, MaxLength, MinLength, Max, Min, ArrayMaxSize, ArrayMinSize } from 'class-validator'";
+    protected static final String IMPORTS = IMPORT_COMMON_VALIDATION_MESSAGES + ";\n" + IMPORT_CLASS_VALIDATOR + ";\n";
```
Comment (lpandzic) [inline infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java]: Those 3 variables make tests really hard to read. I'd rather have duplication with better readability than having to jump all over the place to figure out what the expected value is.

**Review by lpandzic — CHANGES_REQUESTED** [review]

File: infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java (hunk 9 above)
Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java]: You think so? My reasoning with this change was that 
1) when there's less code it's more readable. 
2) these imports were already getting passed to the `TestBase` constructor inside `ClassValidatorDecoratorTestBase` so the test does not test these values, they're just part of test framework setup. If they're not part of the test they should not be visible in the test and the test author should ideally not think about it.

Let me know if this argumentation convinced you, I'll bring back the old version if not.

Comment (lpandzic) [reply infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java]: Less code doesn't automatically mean it's more readable - that's a false assumption. There's no strict correlation.
I'm not sure I follow you on 2. but my point is on removed lines in then statements. Now it's a concatenation of those fields. Combination of those concatenation and 3 separate fields makes it really hard to follow what is going on.

Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java]: I've just remembered why I changed it: it's because I added `"@IsOptional("` to `ClassValidatorDecoratorExtension.DEFAULT_VALIDATIONS`, reran the test and the test still passed, even though `@IsOptional` was not in its imports.

Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java]: So then I had to figure out what was going on and I discovered that the imports were actually getting taken from `ClassValidatorDecoratorTestBase`

Comment (msertic1) [reply infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java]: Let me share my point of view, maybe it will help maybe it will not :)
1. I share my point of view here with @lpandzic. To be honest I have nothing to add then just to agree on fact that less code == more readable code is false assumption 

2. @PiotrMachner i think that I am following you, you have extract this part of code mainly for reusability, because you need to configure ClassValidatorDecoratorTestBase witch takes instance of ClassValidatorDecoratorExtension + import fields (as you said this is configuration of test base).
I am following your logic, but here is one but, this project (at least from my experience) is very complex and from time to time it is hard to follow some parts.

Based on that I would also sided with @lpandzic on comment, mainly because when someone else is going to work with this code, tests should be clear as much as possible (and here by clear I mean, not distracting) and scrolling or jumping in this case is distracting, and one can lose focus while being in debug session of already complex code :)

TL;DR; I would revert to fully stating imports, just to make test easy to follow :)

Comment (msertic1) [reply infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java]: Also i have seen now that @PiotrMachner you replied like hour a go but for some reason i did not see comment. I hate git user experience on reviewing code :)
But for problem mentioned in your above comment ->
How it is solved now ? As I can see solution were to detect where test is located and then to configure list of imports in test (or am i missing something) ?

Why am i asking that, because if we need to still locate tests, when we add something in code and manually add changes, we are still at square one, test could pass even if we have change something, we have just "grouped" imports in static variables, meaning that beside needing to jump from test to static to global variables we did not get any benefits ? (or am i missing something)

Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/test/java/com/infobip/typescript/validation/ClassValidatorDecoratorTestBase.java]: @lpandzic @msertic1 I've brought back the old version (with the full imports), adding only the new import for `IsOptional` 👍

File: infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java (hunk 8 above)
Comment (PiotrMachner) [reply infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java:86]: Yes, and the checks in `isHierarchicalDecoratorNeeded` and `isHierarchyRoot` are the same ones that we had before, so this fix one breakage that my previous changes had introduced in the generates Single Send typescript

**Review by lpandzic — APPROVED** [review]

**Review by msertic1 — APPROVED** [review]
